sensory_jamming.py

```diff
@@ -0,0 +1,342 @@
+        self.callsources = {}
+            # empty dictionary. Will contain the sources of each calls, 
+            # emitted by every agent throughout the simulation.
+        self.boxsize = boxsize 
+            # list of 2 integers. Area (rectangle) within which the agents can 
+            # move, defined by the lengths of edges in meters. 
+            # Has to be of the form [x,y], which is asserted below:
+        
+        assert isinstance(self.boxsize, list) and len(self.boxsize) == 2, "'boxsize' must be a list of 2 elements." 
+            # returns error if boxsize is not of the right format, i.e. [x,y]
+        
+    def Identification(self):
+        
+        for agent in range(self.popsize):
+            # for each agent in the bat population:
+            self.all_ID[int(agent)] = agent
+                # fill-in a list of all agents ID 
+                # from 0 to n = (population size - 1)
+            
+    def Positions(self):
+        
+        for ID in self.all_ID:
+            # for each agent in the bat population:
+           self.agent_initx = rd.uniform(0, self.boxsize[0])
+               # give each agent a random coordinate on the x-axis, 
+               # within the environment boundaries
+           self.agent_inity = rd.uniform(0, self.boxsize[1])
+               # give each agent a random coordinate on the y-axis, 
+               # within the environment boundaries
+           self.all_initpos[int(ID),0] = self.agent_initx
+               # store the x-coordinate in an array
+           self.all_initpos[int(ID),1] = self.agent_inity
+               # store the y-coordinate in an array
+    
+    def Timeline(self):
+            
+        for iteration in range(self.simduration):
+            # for each time step in the simulation:
+            self.realtime += float(self.d_t)
+                # increment timecount with time resolution. 
+            self.timeclock[int(iteration)] = self.realtime
+                # store the new, incremented time count in timeclock.
+
+    class Bat_Jamming_00:
+        # Class implementation nested into Launcher class implementation.
+        # Sets an agent's movement, calling and hearing rules.
+    
+        def __init__(self, ID, movdirection, flightspeed, IPI, max_hear_dist):
+            self.ID = int(ID)
+                # integer. Identification number of the focal agent.
+            self.boxsize = env.boxsize
+                # takes boxsize from env (class Launcher).
+            self.d_t = env.d_t
+                # takes d_t from env (class Launcher).
+            self.movdirection = float(movdirection)
+                # float. Direction (in radians) towards which the agent flies.
+                # Must be between -pi and pi (asserted later).
+            self.stepsize = float(flightspeed * self.d_t) 
+                # float. Distance in meters covered by the agent in 1 time step
+            self.IPI = int(float(IPI * m.pow(10,-3)) / env.d_t)
```

Comment:
here too - it's good to keep code general - keep things in terms of seconds, instead of milliseconds - this allows you to change parameters comfortably to any range you want without causing confusion.